### ISSUE TITLE
Allow turning off tree shaking for external modules or marking external modules as having side-effects

### ISSUE THREAD

**Issue by Lizzy-Gallagher** [webpack/webpack#15486]
## Feature request

**What is the expected behavior?**
In my application, I am seeing various external modules (specifically those with side-effect only imports, i.e. "import "myModule"") being removed by tree shaking from the bundle when in production mode.

According to the [tree shaking documentation](https://webpack.js.org/guides/tree-shaking/), I can mark npm packages as having side effects / not having side-effects, but that same option is not documented for external modules (even when using the function method of identifying externals).

**What is motivation or use case for adding/changing the behavior?**
I would like to preserve these imports since webpack can't really know if I need them or not.

My current workaround is to use "require" on only these imports (since only "import" statements are eligible for treeshaking), but that's hard to remember and I'd love to flag all externals as non-treeshakable.

**How should this be implemented in your opinion?**
Either
1. Change the tree-shaking algorithm to ignore all external modules
2. Provide a new input parameter (or options object) to the callback function for the externals function that allows specifying whether the external has side-effects or not.

**Are you willing to work on this yourself?**
Likely, if this isn't already secretly supported and there's buy-in on which of the two approaches to take.

**Comment by vankop** [comment]
> Change the tree-shaking algorithm to ignore all external modules
all external modules should be preserved in bundle. could you provide a reproducible repo with bug?

**Comment by vankop** [comment]
there is nothing to do from webpack @alexander-akait . user should correctly mark side effects with `sideEffects` flag.

**Comment by Lizzy-Gallagher** [comment]
For clarification, this is specifically for “external modules” which are not loaded from node_modules and may not have a corresponding package.json or even correspond to an npm package! Since the sideEffects flag can only be set in a package.json, this solution is not applicable to the external modules case.

**Comment by alexander-akait** [comment]
@vankop yeah, we don't check `sideEffects` for externals, so we need to allow to set developer `sideEffects` for externals, but maybe we can do it using `module`... need to check

**Comment by steveluscher** [comment]
GitHub didn't create a backlink to our discussion on this, so I'm doing so now, for others to read: https://github.com/webpack/webpack/discussions/16980#discussioncomment-5609953
